File: pybinds_temp/base_scheduler.cpp
#include "base_scheduler.h"
#include "scheduler_outputs.h"
#include "sequence_with_priority.h"
#include "sequence_status.h"
#include <unordered_set>
#include <queue>
#include <vector>
#include <iostream>

using namespace sarathi;

BaseScheduler::BaseScheduler(
    pybind11::object model_config,
    pybind11::object scheduler_config,
    pybind11::object cache_config,
    pybind11::object parallel_config,
    pybind11::object waiting_queue,
    pybind11::object replica_seq_manager,
    pybind11::object metric_store
) :
    metric_store(metric_store),
    model_config(model_config),
    scheduler_config(scheduler_config),
    cache_config(cache_config),
    parallel_config(parallel_config),
    _iteration_id(-1),
    replica_seq_manager(replica_seq_manager),
    seq_seen(std::unordered_set<int>()),
    waiting(),
    num_running_batches(0), 
    new_seqs(std::vector<pybind11::object>()),
    running(std::vector<pybind11::object>())
{
    pybind11::module_ policy_module = pybind11::module_::import("sarathi.core.policy");
    pybind11::object PolicyFactory = policy_module.attr("PolicyFactory");
    this->policy = PolicyFactory.attr("get_policy")("fcfs");

    pybind11::module_ block_manager_module = pybind11::module_::import("sarathi.core.block_space_manager.block_space_manager_registry");
    pybind11::object BlockSpaceManagerRegistry = block_manager_module.attr("BlockSpaceManagerRegistry");
    this->block_manager = BlockSpaceManagerRegistry.attr("get")(
        scheduler_config.attr("get_type")(),
        cache_config.attr("block_size"),
        cache_config.attr("num_gpu_blocks"),
        model_config.attr("max_model_len")
    );

    this->prompt_limit = pybind11::cast<int> (model_config.attr("max_model_len"));  
}

void BaseScheduler::reset_state()
{
    _iteration_id = -1;
}

void BaseScheduler::add_seq(pybind11::object& seq)
{
<<<<<<< HEAD
    float arrived_at = seq.attr("arrived_at")().cast<float>();
    SequenceWithPriority seq_with_priority = SequenceWithPriority(arrived_at, seq);
    waiting.push(seq_with_priority);
=======
    std::cout << "Testing 123" << "\n";
    seq.attr("test")();
    pybind11::object wrapped_seq = seq.attr("create_sequence_with_priority")(seq);
    waiting.attr("put")(wrapped_seq);
>>>>>>> 19deb6a75eef7e3145408ccd7620c385093ef37a
}

bool BaseScheduler::has_unfinished_seqs()
{
    return waiting.size() > 0 || !running.empty();
}

int BaseScheduler::get_num_unfinished_seqs()
{
    return waiting.size() + running.size();
}

// void _schedule() {}

void BaseScheduler::add_to_new_seqs(pybind11::object& seq)
{
    new_seqs.push_back(seq);
}

std::vector<pybind11::object> BaseScheduler::get_new_seqs()
{
    std::vector<pybind11::object> seqs = new_seqs;
    new_seqs.clear();
    return seqs;
}

void BaseScheduler::add_seq_to_seq_manager(pybind11::object& seq)
{
    replica_seq_manager.attr("add_seq")(seq);
}

sarathi::SchedulerOutputs BaseScheduler::schedule()
{
    _iteration_id++;
    std::vector<pybind11::str> ignored_seq_ids = {};
    std::vector<pybind11::str> preempted_seq_ids = {};
    std::vector<pybind11::object> scheduled_seq_metadata_list = {};

    if (num_running_batches > pybind11::cast<int> (parallel_config.attr("pipeline_parallel_size"))) {
        return sarathi::SchedulerOutputs(
            _iteration_id,
            ignored_seq_ids,
            preempted_seq_ids,
            scheduled_seq_metadata_list
        );
    }

    //sarathi::SchedulerOutputs scheduler_outputs = _schedule();

    // if (!scheduler_outputs.is_empty()) {
    //     num_running_batches++;
    // }
    // return scheduler_outputs;

    return sarathi::SchedulerOutputs(
            _iteration_id,
            ignored_seq_ids,
            preempted_seq_ids,
            scheduled_seq_metadata_list
        );

}

void BaseScheduler::free_finished_seqs()
{
    for (pybind11::object& seq: running) {
        if (seq.attr("is_finished")()) {
            seq.attr("_free_seq")();
        }
    }
    std::vector<pybind11::object> new_running;
    for (pybind11::object& seq : running) {
        if (!seq.attr("is_finished")()) {
            new_running.push_back(seq);
        }
    } 
    running = new_running;
}

void BaseScheduler::on_step_completed()
{
    free_finished_seqs();
    num_running_batches--;
}

void BaseScheduler::_allocate(pybind11::object& seq)
{
    block_manager.attr("allocate")(seq);
}

void BaseScheduler::_free_seq(pybind11::object& seq)
{
    block_manager.attr("free")(seq);
}

void BaseScheduler::_append_slot(pybind11::object& seq)
{
    //ASSERT(pybind11::cast<bool> (seq.attr("is_executing")()));
    block_manager.attr("append_slot")(seq);
}

void BaseScheduler::_preempt(pybind11::object& seq)
{
    //ASSERT(pybind11::cast<int> (seq.attr("is_executing")()));
    _free_seq(seq);

}

bool BaseScheduler::_check_request_prompt_length(pybind11::object& seq)
{
    if (pybind11::cast<int> (seq.attr("get_len")()) > prompt_limit) {
        seq.attr("set_status")(sarathi::SequenceStatus::Status::FINISHED_STOPPED);
        if (!waiting.empty()) {
            waiting.pop();
        }

        return false;
    }
    return true;
}

File: sarathi/core/scheduler/base_scheduler.py
from abc import ABC, abstractmethod
from typing import List
from queue import PriorityQueue

from sarathi.config import BaseSchedulerConfig, CacheConfig, ModelConfig, ParallelConfig
from sarathi.core.block_space_manager.block_space_manager_registry import (
    BlockSpaceManagerRegistry,
)
from sarathi.core.datatypes.scheduler_output import SchedulerOutputs
from sarathi.core.datatypes.sequence import Sequence, SequenceStatus, SequenceWithPriority
from sarathi.core.sequence_manager.engine_sequence_manager import EngineSequenceManager
from sarathi.core.policy import PolicyFactory
from sarathi.logger import init_logger
from sarathi.metrics.metrics_store import MetricsStore
from sarathi.utils.threading_utils import synchronized




logger = init_logger(__name__)


class BaseScheduler(ABC):

    def __init__(
        self,
        model_config: ModelConfig,
        scheduler_config: BaseSchedulerConfig,
        cache_config: CacheConfig,
        parallel_config: ParallelConfig,
        waiting_queue : PriorityQueue,
        replica_seq_manager : EngineSequenceManager,
        metric_store : MetricsStore,
    ) -> None:
        self.metrics_store = MetricsStore.get_instance()
        self.model_config = model_config
        self.scheduler_config = scheduler_config
        self.cache_config = cache_config
        self.parallel_config = parallel_config

        # we maintain this just for logging purposes
        self._iteration_id = -1

        # Instantiate the scheduling policy.
        self.policy = PolicyFactory.get_policy(policy_name="fcfs")
        # Create the block space manager.
        self.block_manager = BlockSpaceManagerRegistry.get(
            scheduler_config.get_type(),
            cache_config.block_size,
            cache_config.num_gpu_blocks,
            model_config.max_model_len,
        )
        self.prompt_limit = model_config.max_model_len
        self.replica_seq_manager = replica_seq_manager
        self.new_seqs: List[Sequence] = []
        self.metrics_store = metric_store
        self.seq_seen = set()

        # number of running batches should be less than or equal to the number of pipeline stages
        self.num_running_batches = 0

        # TODO(zhuohan): Use deque instead of list for better performance.
        # Sequence groups in the WAITING state.
        self.waiting : PriorityQueue = PriorityQueue()
        # self.waiting : PriorityQueue = waiting_queue
        # Sequence groups in the RUNNING state.
        self.running: List[Sequence] = []

    def reset_state(self) -> None:
        self._iteration_id = -1

    def add_seq(self, seq: Sequence) -> None:
        # Add sequence groups to the waiting queue. 
        wrapped_seq = SequenceWithPriority(seq.arrived_at, seq)

        self.waiting.put(wrapped_seq)

    def has_unfinished_seqs(self) -> bool:
        return self.waiting.qsize() > 0 or self.running

    def get_num_unfinished_seqs(self) -> int:
        return self.waiting.qsize() + len(self.running)

    @abstractmethod
    def _schedule(self) -> SchedulerOutputs:
        pass

    @synchronized
    def add_to_new_seqs(self, seq: Sequence) -> None:
        self.new_seqs.append(seq)

    @synchronized
    def get_new_seqs(
        self,
    ) -> List[Sequence]:
        new_seqs = self.new_seqs
        self.new_seqs = []
        return new_seqs
    
    @synchronized
    def add_seq_to_seq_manager(self, seq: Sequence) -> None:
        self.replica_seq_manager.add_seq(seq)

    def schedule(self) -> SchedulerOutputs:
        # Schedule sequence groups.
        # This function call changes the internal states of the scheduler
        # such as self.running and self.waiting.
        self._iteration_id += 1

        if self.num_running_batches >= self.parallel_config.pipeline_parallel_size:
            return SchedulerOutputs(
                self._iteration_id,
                ignored_seq_ids=[],
                preempted_seq_ids=[],
                scheduled_seq_metadata_list=[],
            )
        
        scheduler_outputs = self._schedule()

        if not scheduler_outputs.is_empty():
            self.num_running_batches += 1

        return scheduler_outputs

    def free_finished_seqs(self) -> None:
        for seq in self.running:
            if seq.is_finished():
                self._free_seq(seq)
        self.running = [seq for seq in self.running if not seq.is_finished()]

    def on_step_completed(self) -> None:
        self.free_finished_seqs()
        self.num_running_batches -= 1

    def _allocate(self, seq: Sequence) -> None:
        self.block_manager.allocate(seq)

    def _free_seq(self, seq: Sequence) -> None:
        self.block_manager.free(seq)

    def _append_slot(
        self,
        seq: Sequence,
    ) -> None:
        assert seq.is_executing()
        self.block_manager.append_slot(seq)

    def _preempt(
        self,
        seq: Sequence,
    ) -> None:
        assert seq.is_executing()
        self._free_seq(seq)

        wrapped_seq = SequenceWithPriority(seq.arrived_at, seq)
        
        self.waiting.put(wrapped_seq)

    def _check_request_prompt_length(self, seq: Sequence) -> bool:
        if seq.get_len() > self.prompt_limit:
            logger.warning(
                f"Input prompt ({seq.get_len()} tokens) is too long"
                f" and exceeds limit of {self.prompt_limit}"
            )
            seq.set_status(SequenceStatus.FINISHED_IGNORED)
            self.waiting.get(block=False)
            return False

        return True